src/app/clusters/device-energy-management-server/device-energy-management-server.cpp

```diff
@@ -665,15 +847,47 @@ void Instance::HandleRequestConstraintBasedForecast(HandlerContext & ctx,
 
 void Instance::HandleCancelRequest(HandlerContext & ctx, const Commands::CancelRequest::DecodableType & commandData)
 {
-    Status status;
+    Status status                                               = Status::Failure;
+    DataModel::Nullable<Structs::ForecastStruct::Type> forecast = mDelegate.GetForecast();
+
+    if (forecast.IsNull())
+    {
+        ChipLogDetail(AppServer, "Cancelling on a Null forecast!");
+        status = Status::Failure;
+    }
+    else if (forecast.Value().forecastUpdateReason == ForecastUpdateReasonEnum::kInternalOptimization)
+    {
+        ChipLogDetail(AppServer, "Bad Cancel when ESA ForecastUpdateReason was already Internal Optimization!");
+        status = Status::InvalidInState;
+    }
+    else
+    {
+        status = mDelegate.CancelRequest();
+    }
 
-    status = mDelegate.CancelRequest();
     ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
-    if (status != Status::Success)
+}
+
+Status Instance::GetCurrentUtcTime(uint32_t & currentUtcTime) const
```

Comment:
Maybe worth calling this GetCurrentUtcTimeEpoch or just GetCurrentTimeEpochS since I think EPOCH is by definition UTC to be clearer that it is using Matter Epoch time.

Reply:
Function renamed.